lib/dry/types/compiler.rb

```diff
@@ -16,6 +16,11 @@ def visit(node)
         send(:"visit_#{ type }", body)
       end
 
+      def visit_constrained(node)
+        definition, rule, meta = node
+        Types::Constrained.new(visit(definition), rule: visit(rule)).meta(meta)
```

Comment:
It's better to do `Types.rule_compiler.(rule)` and remove `visit_or` and `visit_and`